test/csuite/random_directio/main.c

```diff
@@ -0,0 +1,1127 @@
+	testutil_check(rev->close(rev));
+	testutil_check(session->close(session, NULL));
+	testutil_check(conn->close(conn, NULL));
+
+	for (th = 0; th < nth; th++)
+		free(large_arr[th]);
+	free(large_arr);
+	free(keybuf);
+	free(lastid);
+	return (true);
+}
+
+/*
+ * handler --
+ *	Child signal handler
+ */
+static void
+handler(int sig)
+{
+	pid_t pid;
+	int status, termsig;
+
+	WT_UNUSED(sig);
+	pid = waitpid(-1, &status, WNOHANG|WUNTRACED);
+	if (pid == 0)
+		return;		/* Nothing to wait for. */
+	if (WIFSTOPPED(status))
+		return;
+	if (WIFSIGNALED(status)) {
+		termsig = WTERMSIG(status);
+		if (termsig == SIGCONT || termsig == SIGSTOP)
+			return;
+		printf("Child got signal %d (status = %d, 0x%x)\n",
+		    termsig, status, (unsigned int)status);
+#ifdef WCOREDUMP
+		if (WCOREDUMP(status))
+			printf("Child process id=%d created core file\n", pid);
+#endif
+	}
+
+	/*
+	 * The core file will indicate why the child exited. Choose EINVAL here.
+	 */
+	testutil_die(EINVAL,
+	    "Child process %" PRIu64 " abnormally exited, status=%d (0x%x)",
+	    (uint64_t)pid, status, status);
+}
+
+/*
+ * sleep_wait --
+ *	Wait for a process up to a number of seconds.
+ */
+static void
+sleep_wait(uint32_t seconds, pid_t pid)
+{
+	pid_t got;
+	int status;
+
+	while (seconds > 0) {
+		if ((got = waitpid(pid, &status, WNOHANG|WUNTRACED)) == pid) {
+			if (WIFEXITED(status))
+				testutil_die(EINVAL,
+				    "Child process %" PRIu64 " exited early"
+				    " with status %d", (uint64_t)pid,
+				    WEXITSTATUS(status));
+			if (WIFSIGNALED(status))
+				testutil_die(EINVAL,
+				    "Child process %" PRIu64 " terminated "
+				    " with signal %d", (uint64_t)pid,
+				    WTERMSIG(status));
+		} else if (got == -1)
+			testutil_die(errno, "waitpid");
+
+		--seconds;
+		sleep(1);
+	}
+}
+
+/*
+ * has_direct_io --
+ *	Check for direct I/O support.
+ */
+static bool
+has_direct_io(void)
+{
+#ifdef O_DIRECT
+	return (true);
+#else
+	return (false);
+#endif
+}
+
+/*
+ * main --
+ *	Top level test.
+ */
+int
+main(int argc, char *argv[])
+{
+	struct sigaction sa;
+	struct stat sb;
+	WT_RAND_STATE rnd;
+	pid_t pid;
+	uint32_t datasize, i, interval, ncycles, nth, timeout;
+	int ch, status;
+	const char *method, *working_dir;
+	char buf[1024];
+	bool populate_only, rand_th, rand_time, schema_test, verify_only;
+
+	(void)testutil_set_progname(argv);
+
+	datasize = DEFAULT_DATA_SIZE;
+	nth = MIN_TH;
+	ncycles = DEFAULT_CYCLES;
+	rand_th = rand_time = true;
+	timeout = MIN_TIME;
+	interval = DEFAULT_INTERVAL;
+	populate_only = schema_test = verify_only = false;
+	working_dir = "WT_TEST.random-directio";
+	method = "none";
+	pid = 0;
+
+	if (!has_direct_io()) {
+		fprintf(stderr, "**** test_random_directio: this system does "
+		    "not support direct I/O.\n**** Skipping test.\n");
+		return (EXIT_SUCCESS);
+	}
+	while ((ch = __wt_getopt(progname, argc, argv,
+	    "dh:i:m:n:pST:t:v")) != EOF)
+		switch (ch) {
+		case 'd':
+			datasize = (uint32_t)atoi(__wt_optarg);
```

Comment:
Providing an "-d 50" input argument causes a segmentation fault. I saw that __wt_optarg was NULL.